src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs

```diff
@@ -1259,54 +957,65 @@ protected override int ExecuteCmd(DocumentsOperationContext context)
+
+                        switch (item)
                         {
-                            Debug.Assert(item.Flags.Contain(DocumentFlags.Artificial) == false);
+                            case AttachmentReplicationItem attachment:
+                                item.ToDispose(DocumentIdWorker.GetLowerIdSliceAndStorageKey(context, attachment.Name,out _, out Slice attachmentName));
+                                item.ToDispose(DocumentIdWorker.GetLowerIdSliceAndStorageKey(context, attachment.ContentType, out _, out Slice contentType));
 
-                            var rcvdChangeVector = item.ChangeVector;
+                                database.DocumentsStorage.AttachmentsStorage.PutDirect(context, attachment.Key, attachmentName, contentType, attachment.Base64Hash,
+                                    attachment.ChangeVector);
 
-                            context.LastDatabaseChangeVector = ChangeVectorUtils.MergeVectors(item.ChangeVector, context.LastDatabaseChangeVector);
-                            if (item.Type == ReplicationBatchItem.ReplicationItemType.Attachment)
-                            {
-                                database.DocumentsStorage.AttachmentsStorage.PutDirect(context, item.Key, item.Name,
-                                    item.ContentType, item.Base64Hash, item.ChangeVector);
+                                if (_replicationInfo.ReplicatedAttachmentStreams.TryGetValue(attachment.Base64Hash, out var attachmentWithStream))
+                                {
+                                    Debug.Assert(SliceComparer.Compare(attachmentWithStream.Base64Hash, attachment.Base64Hash) == 0);
+                                    database.DocumentsStorage.AttachmentsStorage.PutAttachmentStream(context, attachment.Key, attachmentWithStream.Base64Hash,
+                                        attachmentWithStream.Stream);
+                                    _replicationInfo.ReplicatedAttachmentStreams.Remove(attachment.Base64Hash);
+                                }
 
-                                if (_replicationInfo.ReplicatedAttachmentStreams.TryGetValue(item.Base64Hash, out var attachmentStream))
+                                break;
+                            case AttachmentTombstoneReplicationItem attachmentTombstone:
+                                database.DocumentsStorage.AttachmentsStorage.DeleteAttachmentDirect(context, attachmentTombstone.Key, false, "$fromReplication", null,
+                                    rcvdChangeVector,
+                                    attachmentTombstone.LastModifiedTicks);
+                                break;
+
+                            case RevisionTombstoneReplicationItem revisionTombstone:
+                                item.ToDispose(Slice.From(context.Allocator, revisionTombstone.Id, out var id)); // TODO: From not required
```

Comment:
I don't understand the `TODO`?

Reply:
`Slice.From` will allocate and copy the value of `revisionTombstone.Id`. This is not required here because we (the current context) already own `revisionTombstone.Id`.

This is used only to convert from `LazyStringValue` to `Slice`.